### DISCUSSION TITLE
How Users can contribute and improve Argos Translate

### DISCUSSION THREAD

**Discussion by fossdd** [argosopentech/argos-translate#61] [Ideas]
I currently search a option how end-users can improve the AI. Like Google has a interface with badges, etc.

The usual end-user doesn't know much about training-scripts, models, etc. so a soulution where users can fix sentences in a web UI would nice.

I don't know how to implement it, but it would nice, so that the world can share their language experience to improve Open Source Translation.

**Comment by PJ-Finlay** [comment]
The Argos Translate and LibreTranslate GUIs don't currently support this but you could save manual translations as JSON:
```
[
  {
    "lang_code": "en",
    "author": "William Shakespeare"
    "value": "To be, or not to be, that is the question"
  },
  {
    "lang_code": "es",
    "author": "P.J. Finlay",
    "value": "Ser, o no ser, esa es la cuestión."
  }
]
```

**Reply by fossdd** [reply]
Ah nice. But such a UI would nice.

**Reply by PJ-Finlay** [reply]
There's now a git repo if anyone's interested in contributing translations:

- [Argos Parallel Corpus](https://github.com/argosopentech/argos-parallel-corpus)

**Reply by philhoefer** [reply]
Thanks for opening the repo. I'm thinking of contributing. I do however have some questions:

1. The given example is a list/array of objects. Shouldn't it rather be a list of lists of objects, so translations of the same sentence can be grouped together?
2. What should the "author" data mean? It doesn't differentiate between original authors and translators. Is it even helpful? If the corpus is supposed to be licensed as CC0, it should only contain sentences that aren't copyrightable or are public domain (or similar) and thus attribution wouldn't be necessary. I think maybe a more meaningful alternative to the "author" data would be a "contributor" field which could give people some additional incentive to contribute (some might be motivated by a "contributor high score table").
3. How could/would the corpus data be used to improve the AI, especially in regard to existing larger corpuses.
4. I'm thinking about building a tool that lets users rate translations. Would such data be helpful?

**Reply by PJ-Finlay** [reply]
> The given example is a list/array of objects. Shouldn't it rather be a list of lists of objects, so translations of the same sentence can be grouped together?

Yes, good point.

> What should the "author" data mean? It doesn't differentiate between original authors and translators. Is it even helpful? If the corpus is supposed to be licensed as CC0, it should only contain sentences that aren't copyrightable or are public domain (or similar) and thus attribution wouldn't be necessary. I think maybe a more meaningful alternative to the "author" data would be a "contributor" field which could give people some additional incentive to contribute (some might be motivated by a "contributor high score table").

I would consider the author field optional and refer to the person who wrote that specific text (original composition or translator). Certainly not necessary but in the future the information could be valuable if, for example, you wanted to up/down weight certain authors to affect writing style. 

> How could/would the corpus data be used to improve the AI, especially in regard to existing larger corpuses.

In the near term with manually collected data probably wouldn't move the needle much on languages with lots of existing data. If there is a type of translation that we're consistently failing at (maybe a specific phrase) adding it may fix the issue. For languages with less data and using some mild automation I think collecting our own data could also be useful.

> I'm thinking about building a tool that lets users rate translations. Would such data be helpful?

This would seem like valuable data to gather, for now we could probably use the 5 star translations as data to train new models. I think the ideal situation would be to be able to have LibreTranslate web app users rate translations and propose their own.

Glad to hear you're interested in contributing, I just quickly put this up because there seemed to be interest if you think there's a better way to do something you're probably right!

**Comment by ghost** [comment]
Could [Dialect](https://github.com/dialect-app/dialect) be used as the GUI which submits translation improvements?

https://github.com/dialect-app/dialect/issues/194

Thank you

**Reply by PJ-Finlay** [reply]
https://github.com/argosopentech/argos-parallel-corpus is currently essentially an empty repo, if you want to add an integration I can accept a pull request.

**Reply by ghost** [reply]
### 1. Will it enable definitions for the words?

![kép](https://user-images.githubusercontent.com/17465872/132155769-5fbade42-5ace-43a2-a01e-466b820abd0c.png)

https://szotar.sztaki.hu/en/search?fromlang=eng&tolang=hun&fromlang=eng&tolang=hun&searchWord=charge&langcode=en&u=0&langprefix=en%2F&searchMode=WORD_PREFIX&viewMode=full&ignoreAccents=0

### 2. Will it enable pronunciation?

For example, [minute](https://www.thefreedictionary.com/minute) spelt the same but pronounced differently has multiple definitions:
a) 60 seconds
b) very small

### 3. Homographs, homonyms, heteronyms

![kép](https://user-images.githubusercontent.com/17465872/132156376-fd3ff8d7-cf71-4aaf-82cc-3cc891a1b978.png)

Thank you

**Reply by PJ-Finlay** [reply]
If Dialect supports it pronunciations could be useful

**Reply by ghost** [reply]
Perhaps @fushinari can better answer your question?

**Reply by mufeedali** [reply]
Yes, it does :)

See: https://github.com/dialect-app/dialect/pull/109 and https://github.com/dialect-app/dialect/pull/148

**Comment by EmanuelLoos** [comment]
You might also want to take a look at how DeepL does it. It is the best translation service I know of (non-free, as in freedom) and I suppose that is because DeepL feeds its AI with user input (you can read it in their privacy policy) without the user even noticing it that much. I also started a discussion about implementing something similar into LibreTranslate [here](https://community.libretranslate.com/t/be-inspired-by-features-of-other-online-translation-services-and-discuss-reverse-engeneering-them/44). What I have in mind would be more privacy friendly and noticeable than what DeepL does though: I suggest reverse engineering the way DeepL collects data but storing them on the computer of the user only at first and asking the user about each correction if it should be contributed when the user finished translating.

**Comment by PJ-Finlay** [comment]
See current work in LibreTranslate: https://github.com/LibreTranslate/LibreTranslate/pull/151
